File: Java/OpenTraceability/src/main/opentraceability/models/events/associationevent.kt
package models.events

import interfaces.IEvent
import interfaces.IEventKDE
import java.util.*
import models.identifiers.*
import models.events.kdes.CertificationList
import models.events.*
import models.identifiers.EPC
import models.identifiers.PGLN
import utility.attributes.*
import utility.attributes.OpenTraceabilityXmlIgnoreAttribute
import java.time.Duration
import java.lang.reflect.Type
import java.net.URI
import java.time.OffsetDateTime

class AssociationEvent : EventBase(), IEvent {

    @OpenTraceabilityAttribute("","parentID", 7)
    var ParentID: EPC? = null

    @OpenTraceabilityProductsAttribute("childQuantityList", EPCISVersion.V2, EventProductType.Child, 9, OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute("childEPCs", EventProductType.Child, 8, OpenTraceabilityProductsListType.EPCList)
    var Children: ArrayList<EventProduct> = ArrayList<EventProduct>()

    @OpenTraceabilityAttribute("","action", 10)
    override  var Action: EventAction? = null

    @OpenTraceabilityAttribute("","bizStep", 11)
    override  var BusinessStep: URI? = null


    @OpenTraceabilityAttribute("","disposition", 12)
    override  var Disposition: URI? = null


    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","readPoint", 13)
    override  var ReadPoint: EventReadPoint = EventReadPoint()


    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","bizLocation", 14)
    override lateinit var Location: EventLocation

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("bizTransaction")
    @OpenTraceabilityAttribute("","bizTransactionList", 15, EPCISVersion.V2)
    override  var BizTransactionList: ArrayList<EventBusinessTransaction> = ArrayList<EventBusinessTransaction>()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("source")
    @OpenTraceabilityAttribute("","sourceList", 16, EPCISVersion.V2)
    override var SourceList: ArrayList<EventSource> = ArrayList<EventSource>()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("destination")
    @OpenTraceabilityAttribute("","destinationList", 17, EPCISVersion.V2)
    override var DestinationList: ArrayList<EventDestination> = ArrayList<EventDestination>()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("sensorElement")
    @OpenTraceabilityAttribute("","sensorElementList", 18, EPCISVersion.V2)
    override var SensorElementList: ArrayList<SensorElement> = ArrayList<SensorElement>()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","persistentDisposition", 19)
    override var PersistentDisposition: PersistentDisposition? = PersistentDisposition()

    var ILMD: EventILMD = EventILMD()

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("","type", 0)
    override lateinit var EventType: EventType



}

File: CSharp/OpenTraceability/Models/Events/AssociationEvent.cs
﻿using OpenTraceability.Interfaces;
using OpenTraceability.Models.Identifiers;
using OpenTraceability.Utility.Attributes;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace OpenTraceability.Models.Events
{
    public class AssociationEvent : EventBase, IEvent
    {
        [OpenTraceability("parentID", 7)]
        public EPC ParentID { get; set; }

        [OpenTraceabilityProducts("childQuantityList", EPCISVersion.V2, EventProductType.Child, 9, OpenTraceabilityProductsListType.QuantityList)]
        [OpenTraceabilityProducts("childEPCs", EventProductType.Child, 8, OpenTraceabilityProductsListType.EPCList)]
        public List<EventProduct> Children { get; set; } = new List<EventProduct>();

        [OpenTraceability("action", 10)]
        public EventAction? Action { get; set; }

        [OpenTraceability("bizStep", 11)]
        public Uri BusinessStep { get; set; }

        [OpenTraceability("disposition", 12)]
        public Uri Disposition { get; set; }

        [OpenTraceabilityObject]
        [OpenTraceability("readPoint", 13)]
        public EventReadPoint ReadPoint { get; set; }

        [OpenTraceabilityObject]
        [OpenTraceability("bizLocation", 14)]
        public EventLocation Location { get; set; }

        [OpenTraceabilityObject]
        [OpenTraceabilityArray("bizTransaction")]
        [OpenTraceability("bizTransactionList", 15, EPCISVersion.V2)]
        public List<EventBusinessTransaction> BizTransactionList { get; set; } = new List<EventBusinessTransaction>();

        [OpenTraceabilityObject]
        [OpenTraceabilityArray("source")]
        [OpenTraceability("sourceList", 16, EPCISVersion.V2)]
        public List<EventSource> SourceList { get; set; } = new List<EventSource>();

        [OpenTraceabilityObject]
        [OpenTraceabilityArray("destination")]
        [OpenTraceability("destinationList", 17, EPCISVersion.V2)]
        public List<EventDestination> DestinationList { get; set; } = new List<EventDestination>();

        [OpenTraceabilityObject]
        [OpenTraceabilityArray("sensorElement")]
        [OpenTraceability("sensorElementList", 18, EPCISVersion.V2)]
        public List<SensorElement> SensorElementList { get; set; } = new List<SensorElement>();

        [OpenTraceabilityObject]
        [OpenTraceability("persistentDisposition", 19)]
        public PersistentDisposition PersistentDisposition { get; set; }

        public EventILMD ILMD { get => null; }


        [OpenTraceabilityXmlIgnore]
        [OpenTraceability("type", 0)]
        public EventType EventType => EventType.AssociationEvent;

        public EventILMD GetILMD() => ILMD;

        public ReadOnlyCollection<EventProduct> Products
        {
            get
            {
                List<EventProduct> products = new List<EventProduct>();
                if (this.ParentID != null)
                {
                    products.Add(new EventProduct(this.ParentID)
                    {
                        Type = EventProductType.Parent
                    });
                }                
                products.AddRange(this.Children);
                return new ReadOnlyCollection<EventProduct>(products);
            }
        }

        public void AddProduct(EventProduct product)
        {
            if (product.Type == EventProductType.Parent)
            {
                if (product.Quantity != null)
                {
                    throw new Exception("Parents do not support quantity.");
                }
                this.ParentID = product.EPC;
            }
            else if (product.Type == EventProductType.Child)
            {
                this.Children.Add(product);
            }
            else
            {
                throw new Exception("Association event only supports children and parents.");
            }
        }

        public void RemoveProduct(EventProduct product)
        {
            if (product.Type == EventProductType.Parent)
            {
                if (product.Quantity != null)
                {
                    throw new Exception("Parents do not support quantity.");
                }
                this.ParentID = null;
            }
            else if (product.Type == EventProductType.Child)
            {
                this.Children.Remove(product);
            }
            else
            {
                throw new Exception($"Association event only supports children and parents and does not contain this product as either one: {product.EPC}");
            }
        }
    }
}